### PR TITLE
Add explicit error if SCF is used with no kinetic term

### PR DESCRIPTION
**Pull request by epolack**

### CODE REVIEW

**Comment by antoine-levitt** [comment]
Technically it's OK to do it (just the eigensolver will do weird stuff), it's just the preconditioner that needs it, so put that check there probably?

**Comment by epolack** [comment]
With `diag_full`, even with `scaling` arbitrarily put to `1`, it still crashes at line `Afull = Hermitian(Array(A))`.

**Comment by antoine-levitt** [comment]
You mean if you've got no hamiltonian contribution at all? OK yeah but in that case you deserve what happens to you :-p I don't think we should have a check for that level of craziness. So let's just have a check that there is indeed a kinetic term in the preconditioner, and then we just have a crash when there's no hamiltonian at all

File: src/eigen/preconditioners.jl [hunk 1]
```diff
@@ -33,7 +33,13 @@ mutable struct PreconditionerTPA{T <: Real}
...
 
 function PreconditionerTPA(basis::PlaneWaveBasis{T}, kpt::Kpoint; default_shift=1) where T
-    scaling = only([t for t in basis.model.term_types if t isa Kinetic]).scaling_factor
+    kinetic_term = [t for t in basis.model.term_types if t isa Kinetic]
+    if length(kinetic_term) >= 1
+        scaling = only(kinetic_term).scaling_factor
+    else
+        # If user do strange computations without a kinetic term.
+        scaling = 1
+    end
```
Comment (antoine-levitt) [inline src/eigen/preconditioners.jl:42]: This is a type instability. But here just raise an error; the preconditioner is not relevant in this case, users should disable preconditioning. So just `kinetic_terms = []; length() >= 1 || error(); scaling = only(kinetic_terms).scaling`

**Review by antoine-levitt — COMMENTED** [review]
no linebreak and I'm happy :-)

File: src/eigen/preconditioners.jl [hunk 2]
```diff
@@ -33,7 +33,9 @@ mutable struct PreconditionerTPA{T <: Real}
 end
 
 function PreconditionerTPA(basis::PlaneWaveBasis{T}, kpt::Kpoint; default_shift=1) where T
-    scaling = only([t for t in basis.model.term_types if t isa Kinetic]).scaling_factor
+    kinetic_term = [t for t in basis.model.term_types if t isa Kinetic]
+    length(kinetic_term) >= 1 || error("Preconditioner should be disabled when no kinetic term is used.")
```
Comment (mfherbst) [inline src/eigen/preconditioners.jl:37]:
```suggestion
    isempty(kinetic_term) && error("Preconditioner should be disabled when no Kinetic term is used.")
```

**Review by mfherbst — COMMENTED** [review]
To make me happy :)

**Pull request merged**